File: backend/application/config/configuration_provider.py
"""Configuration Provider (Phase 14.3.2).

Thread-safe production configuration provider runtime coordinating configuration state,
sources, context, capabilities, health reporting, statistics, and diagnostics snapshots.
"""

from datetime import datetime, timezone
import logging
from threading import RLock
from typing import Dict, Optional

from backend.application.config.configuration_source_manager import ConfigurationSourceManager
from backend.application.config.interfaces import IConfigurationProvider
from backend.application.config.models import (
    ConfigurationCapabilities,
    ConfigurationContext,
    ConfigurationDiagnostics,
    ConfigurationHealth,
    ConfigurationProfileType,
    ConfigurationRuntimeState,
    ConfigurationStatistics,
)

logger = logging.getLogger(__name__)


class ConfigurationProvider(IConfigurationProvider):
    """Production ConfigurationProvider runtime executing configuration state & source management."""

    def __init__(
        self,
        source_manager: Optional[ConfigurationSourceManager] = None,
        config_context: Optional[ConfigurationContext] = None,
    ) -> None:
        """Initialize ConfigurationProvider using Constructor Dependency Injection.

        Args:
            source_manager: Optional ConfigurationSourceManager instance.
            config_context: Optional ConfigurationContext snapshot.
        """
        self._lock = RLock()
        self._context = config_context or ConfigurationContext()
        self._source_manager = source_manager or ConfigurationSourceManager()
        self._state: ConfigurationRuntimeState = ConfigurationRuntimeState.UNINITIALIZED
        self._reload_count: int = 0

    @property
    def source_manager(self) -> ConfigurationSourceManager:
        """Get the underlying ConfigurationSourceManager instance."""
        with self._lock:
            return self._source_manager

    def initialize(self) -> ConfigurationRuntimeState:
        """Initialize provider runtime state to READY."""
        with self._lock:
            if self._state == ConfigurationRuntimeState.READY:
                return self._state
            logger.info("Initializing ConfigurationProvider for environment '%s'...", self._context.environment.value)
            self._state = ConfigurationRuntimeState.INITIALIZING
            self._state = ConfigurationRuntimeState.READY
            logger.info("ConfigurationProvider initialized successfully. State -> READY.")
            return self._state

    def shutdown(self) -> ConfigurationRuntimeState:
        """Shutdown provider runtime operations to STOPPED."""
        with self._lock:
            if self._state == ConfigurationRuntimeState.STOPPED:
                return self._state
            logger.info("Shutting down ConfigurationProvider...")
            self._state = ConfigurationRuntimeState.STOPPING
            self._state = ConfigurationRuntimeState.STOPPED
            logger.info("ConfigurationProvider shutdown complete. State -> STOPPED.")
            return self._state

    def restart(self) -> ConfigurationRuntimeState:
        """Restart provider runtime operations."""
        with self._lock:
            logger.info("Restarting ConfigurationProvider...")
            self.shutdown()
            return self.initialize()

    def health(self) -> ConfigurationHealth:
        """Get health assessment snapshot of the provider and source manager."""
        with self._lock:
            sm_health = self._source_manager.health()
            is_healthy = self._state in (
                ConfigurationRuntimeState.READY,
                ConfigurationRuntimeState.INITIALIZING,
                ConfigurationRuntimeState.UNINITIALIZED,
            ) and sm_health.is_healthy
            issues = sm_health.issues if is_healthy else (f"ConfigurationProvider state is {self._state.value}.",) + sm_health.issues
            return ConfigurationHealth(
                is_healthy=is_healthy,
                state=self._state,
                issues=issues,
                checked_at=datetime.now(timezone.utc),
            )

    def statistics(self) -> ConfigurationStatistics:
        """Get statistics metrics snapshot of the provider and sources."""
        with self._lock:
            sm_stats = self._source_manager.statistics()
            metrics = {
                "reload_count": float(self._reload_count),
            }
            metrics.update(sm_stats.metrics)
            return ConfigurationStatistics(
                total_properties_loaded=sm_stats.total_properties_loaded,
                active_sources_count=sm_stats.active_sources_count,
                profiles_loaded_count=1,
                reload_count=self._reload_count,
                metrics=metrics,
            )

    def capabilities(self) -> ConfigurationCapabilities:
        """Get capability definitions of the provider."""
        return ConfigurationCapabilities(
            supports_dotenv=True,
            supports_json=True,
            supports_yaml=True,
            supports_environment_override=True,
            supports_remote_sources=True,
            supports_hot_reload=True,
            supports_secret_masking=True,
            supports_type_casting=True,
        )

    def diagnostics(self) -> ConfigurationDiagnostics:
        """Get resolution diagnostics snapshot."""
        with self._lock:
            stats = self.statistics()
            return ConfigurationDiagnostics(
                state=self._state,
                health=self.health(),
                statistics=stats,
                active_profile_name=self._context.environment.value.lower(),
                active_sources_count=stats.active_sources_count,
                metrics=stats.metrics,
                timestamp=datetime.now(timezone.utc),
            )

    def get_context(self) -> ConfigurationContext:
        """Get execution context snapshot."""
        with self._lock:
            return self._context

File: frontend/src/runtime/config/configuration_provider.ts
/**
 * Configuration Provider Implementation (Phase 16.3.2).
 *
 * Implements IConfigurationProvider owning configuration runtime state transitions,
 * telemetry statistics, health evaluation, context metadata, capabilities reporting,
 * and source resolution delegation.
 */

import {
  ConfigurationCapabilities,
  ConfigurationConfiguration,
  ConfigurationContext,
  ConfigurationDiagnostics,
  ConfigurationEntry,
  ConfigurationHealth,
  ConfigurationRuntimeState,
  ConfigurationSnapshot,
  ConfigurationSourceRegistration,
  ConfigurationState,
  ConfigurationStatistics,
  createConfigurationCapabilities,
  createConfigurationConfiguration,
  createConfigurationContext,
  createConfigurationDiagnostics,
  createConfigurationHealth,
  createConfigurationSourceRegistration,
  createConfigurationState,
  createConfigurationStatistics,
} from './models';
import { IConfigurationProvider, IConfigurationSource } from './interfaces';
import { SourceRegistry } from './source_registry';
import { ConfigurationSourceManager } from './configuration_source_manager';
import { MemoryConfigurationSource } from './memory_source';

export class ConfigurationProvider implements IConfigurationProvider {
  private _runtimeState: ConfigurationRuntimeState = ConfigurationRuntimeState.UNINITIALIZED;
  private readonly _config: ConfigurationConfiguration;
  private readonly _capabilities: ConfigurationCapabilities;
  private readonly _context: ConfigurationContext;

  private readonly _registry: SourceRegistry;
  private readonly _sourceManager: ConfigurationSourceManager;
  private readonly _defaultMemorySource: MemoryConfigurationSource;

  private _startedAt: string | null = null;
  private _initializations = 0;
  private _shutdowns = 0;
  private _restarts = 0;
  private _errors = 0;

  constructor(
    config?: ConfigurationConfiguration,
    capabilities?: ConfigurationCapabilities,
    context?: ConfigurationContext,
    registry?: SourceRegistry,
  ) {
    this._config = config ?? createConfigurationConfiguration();
    this._capabilities = capabilities ?? createConfigurationCapabilities();
    this._context = context ?? createConfigurationContext();

    this._registry = registry ?? new SourceRegistry();
    this._sourceManager = new ConfigurationSourceManager(this._registry);

    this._defaultMemorySource = new MemoryConfigurationSource();
    this._registry.register(this._defaultMemorySource);
  }

  public initialize(): ConfigurationHealth {
    if (
      this._runtimeState === ConfigurationRuntimeState.INITIALIZING ||
      this._runtimeState === ConfigurationRuntimeState.READY
    ) {
      return this.health();
    }

    this._runtimeState = ConfigurationRuntimeState.INITIALIZING;
    this._runtimeState = ConfigurationRuntimeState.READY;
    this._startedAt = new Date().toISOString();
    this._initializations++;

    return this.health();
  }

  public shutdown(): ConfigurationHealth {
    if (this._runtimeState === ConfigurationRuntimeState.STOPPED) {
      return this.health();
    }

    this._runtimeState = ConfigurationRuntimeState.STOPPING;
    this._runtimeState = ConfigurationRuntimeState.STOPPED;
    this._startedAt = null;
    this._shutdowns++;

    return this.health();
  }

  public restart(): ConfigurationHealth {
    this._restarts++;
    this.shutdown();
    return this.initialize();
  }

  public health(): ConfigurationHealth {
    const healthy = this._runtimeState === ConfigurationRuntimeState.READY;
    const message = healthy
      ? 'Configuration runtime is ready and operational.'
      : `Configuration runtime is in state ${this._runtimeState}.`;

    return createConfigurationHealth({
      healthy,
      runtimeState: this._runtimeState,
      message,
    });
  }

  public statistics(): ConfigurationStatistics {
    const uptime =
      this._runtimeState === ConfigurationRuntimeState.READY && this._startedAt
        ? Math.max(0, Math.floor((Date.now() - new Date(this._startedAt).getTime()) / 1000))
        : 0;

    return createConfigurationStatistics({
      initializations: this._initializations,
      shutdowns: this._shutdowns,
      restarts: this._restarts,
      errors: this._errors,
      uptime,
    });
  }

  public capabilities(): ConfigurationCapabilities {
    return this._capabilities;
  }

  public diagnostics(): ConfigurationDiagnostics {
    const sourceRegs: ConfigurationSourceRegistration[] = this.listSources().map((s) =>
      createConfigurationSourceRegistration({
        sourceName: s.name,
        priority: s.priority,
        enabled: s.enabled,
      }),
    );

    return createConfigurationDiagnostics({
      health: this.health(),
      statistics: this.statistics(),
      capabilities: this.capabilities(),
      context: this._context,
      sources: sourceRegs,
      snapshot: this.createSnapshot(),
      timestamp: new Date().toISOString(),
    });
  }

  public state(): ConfigurationState {
    return createConfigurationState({
      runtimeState: this._runtimeState,
      initialized: this._runtimeState === ConfigurationRuntimeState.READY,
      startedAt: this._startedAt,
    });
  }

  public configuration(): ConfigurationConfiguration {
    return this._config;
  }

  public context(): ConfigurationContext {
    return this._context;
  }

  public registerSource(source: IConfigurationSource): void {
    this._registry.register(source);
  }

  public unregisterSource(sourceName: string): boolean {
    return this._registry.unregister(sourceName);
  }

  public get<T = unknown>(key: string, defaultValue?: T): T | undefined {
    return this._sourceManager.get<T>(key, defaultValue);
  }

  public has(key: string): boolean {
    return this._sourceManager.has(key);
  }

  public getEntry(key: string): ConfigurationEntry | undefined {
    return this._sourceManager.getEntry(key);
  }

  public getAll(): Readonly<Record<string, unknown>> {
    return this._sourceManager.getAll();
  }

  public createSnapshot(): ConfigurationSnapshot {
    return this._sourceManager.createSnapshot();
  }

  public listSources(): ReadonlyArray<IConfigurationSource> {
    return this._registry.listSources();
  }
}
